scu_builders.py

```diff
@@ -0,0 +1,333 @@
+        found_includes, found_exceptions = find_files_in_folder(
+            main_folder, folders[d], found_includes, extension, sought_exceptions, found_exceptions
+        )
+
+    found_includes = sorted(found_includes)
+
+    # calculate how many lines to write in each file
+    total_lines = len(found_includes)
+
+    # adjust number of output files according to whether DEV or release
+    num_output_files = 1
+    if _is_release_build:
+        # always have a maximum in release
+        includes_per_scu = 8
+        num_output_files = max(math.ceil(total_lines / float(includes_per_scu)), 1)
+    else:
+        if includes_per_scu > 0:
+            num_output_files = max(math.ceil(total_lines / float(includes_per_scu)), 1)
+
+    lines_per_file = math.ceil(total_lines / float(num_output_files))
+    lines_per_file = max(lines_per_file, 1)
+
+    start_line = 0
+    file_number = 0
+
+    # These do not vary throughout the loop
+    output_folder = abs_main_folder + "/scu/"
+    output_filename_prefix = "scu_" + out_filename
+
+    # Clear out any existing files (usually we will be overwriting,
+    # but we want to remove any that are pre-existing that will not be
+    # overwritten, so as to not compile anything stale)
+    clear_out_existing_files(output_folder, extension)
+
+    for file_count in range(0, num_output_files):
+        end_line = start_line + lines_per_file
+
+        # special case to cover rounding error in final file
+        if file_count == (num_output_files - 1):
+            end_line = len(found_includes)
+
+        write_output_file(
+            file_count, found_includes, start_line, end_line, output_folder, output_filename_prefix, extension
+        )
+
+        start_line = end_line
+
+    # Write the exceptions each in their own scu gen file,
+    # so they can effectively compile in "old style / normal build".
+    for exception_count in range(len(found_exceptions)):
+        write_exception_output_file(
+            exception_count, found_exceptions[exception_count], output_folder, output_filename_prefix, extension
+        )
+
+
+def generate_scu_files(verbose, is_release_build):
+
+    print("=============================")
+    print("Single Compilation Unit Build")
+    print("=============================")
+    print("Generating SCU build files")
+    global _verbose
+    _verbose = verbose
+    global _is_release_build
+    _is_release_build = is_release_build
+
+    # check we are running from the correct folder
+    curr_folder = os.path.abspath("./")
+    # parent_path = basename(normpath(curr_folder))
+
+    if folder_not_found("core") or folder_not_found("platform") or folder_not_found("scene"):
+        print("ERROR - scu_builders.py must be run from the godot folder.")
+        return
+
+    process_folder(["core"])
+    process_folder(["core/crypto"])
+    process_folder(["core/debugger"])
+    process_folder(["core/extension"])
+    process_folder(["core/input"])
```

Comment:
Why do we need to list all of those manually, when we already call `add_source_files_scu` directly in each of those SCsubs?

It would be better if we could keep the responsibility to register stuff for the SCU build where the files are added to the build, instead of in a helper script.

Reply:
We could potentially do that, but my aim was to keep everything as simple as possible in the `SCsub` files.

Essentially, for those not using SCU builds, my aim was that the impact should be as minimal as possible so that you e.g. don't need to understand SCU builds in order to editor the `SCsub` files (keep the cognitive load as low as possible). In many cases the only change to the `SCsub` is using the `add_source_files_scu()` function. Adding the unnecessary details into the SCsub is likely to put off people modifying it, and could lead to resentment.

Additionally, most of the maintenance of the SCU build can be kept to `scu_builders.py` (e.g. file exclusions) which keeps the commits / diffs easier to understand.

So we *can* take this approach, but I'm not sure whether it would offer significant advantages in return for the disadvantages.